exonum-java-binding/core/src/main/java/com/exonum/binding/runtime/Pf4jServiceLoader.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.runtime;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.exonum.binding.service.ServiceModule;
+import com.google.common.base.MoreObjects;
+import com.google.inject.Inject;
+import java.net.URI;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.Supplier;
+import org.pf4j.PluginManager;
+import org.pf4j.PluginState;
+
+/**
+ * A loader of services as PF4J plugins. Such plugins are required to have PluginId set in
+ * a certain format ('groupId:artifactId:version'); have a single {@link ServiceModule} as
+ * an extension.
+ *
+ * @see <a href="https://pf4j.org/doc/getting-started.html">PF4J docs</a>
+ */
+final class Pf4jServiceLoader implements ServiceLoader {
+
+  private static final Comparator<ServiceId> SERVICE_ID_COMPARATOR =
+      Comparator.comparing(ServiceId::getGroupId)
+          .thenComparing(ServiceId::getArtifactId)
+          .thenComparing(ServiceId::getVersion);
+
+  private final PluginManager pluginManager;
+  private final SortedMap<ServiceId, LoadedServiceDefinition> loadedServices;
+
+  @Inject
+  Pf4jServiceLoader(PluginManager pluginManager) {
+    this.pluginManager = checkNotNull(pluginManager);
+    loadedServices = new TreeMap<>(SERVICE_ID_COMPARATOR);
+  }
+
+  /**
+   * Loads a service as PF4J artifact.
+   *
+   * <p>Verification steps involve metadata validation, starting the plugin and checking it provides
+   * a single ServiceModule as an extension.
+   */
+  @Override
+  public LoadedServiceDefinition loadService(URI artifactLocation)
+      throws ServiceLoadingException {
+    // Load a plugin
+    String pluginId = loadPlugin(artifactLocation);
+    try {
+      // Start the plugin
+      startPlugin(pluginId);
+
+      // Load the service definition
+      return loadDefinition(pluginId);
+    } catch (IllegalArgumentException e) {
+      pluginManager.unloadPlugin(pluginId);
+      throw new ServiceLoadingException(e);
```

Comment:
let's add error description here

Reply:
:+1: pluginId won't hurt. The details, however, are in the cause (IAE)